File: app/game_commands/scrap_unit.rb
class ScrapUnit < GameCommand
  def initialize(game, user, params)
    super(game, user, params)
    @x, @y = *@params.values_at(:x, :y)
  end

  def execute
    require_current_player
    require_running_game

    base = require_friendly_base(@x, @y)
    unit = require_current_player_unit(@x, @y)

    raise CommandError.new("Unit cannot be scrapped there.") unless base.can_build_unit_type?(unit.unit_type)
    raise CommandError.new("Cannot scrap after attacking.") unless unit.attacks == 0

    destroy_unit(unit)
    modify_credits(@game.player_number_for_user(@user), @game.user_credits(@user) + unit.scrap_value)
  end
end

File: public/javascripts/ec/game/game_commands/scrap_unit.js
var ScrapUnit = new JS.Class(GameCommand, {
  initialize: function(game, user, params) {
    this.callSuper();
    this.x = params.x;
    this.y = params.y;
  },

  class_name: function() {
    return 'ScrapUnit';
  },

  execute: function() {
    var unit = this.game.unit_at(this.x, this.y);
    var pn = this.game.player_number(this.user);

    this.destroy_unit(unit);
    this.modify_credits(pn, this.game._player_credits[pn - 1] + unit.scrap_value());
  },

  can_undo: function() {
    return true;
  }
});
